### PR TITLE
Add local group_by_key operation

### PR DESCRIPTION
**Pull request by zachferr**
## Description
Adding local group_by_key and associated tests

## Affected Dependencies
N/A

## How has this been tested?
- Unit test

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/pipeline_operations.py [hunk 1]
```diff
@@ -156,8 +157,11 @@ def map_tuple(self, col, fn, stage_name: str = None):
     def map_values(self, col, fn, stage_name: str):
         pass
 
-    def group_by_key(self, col, stage_name: str):
-        pass
+    def group_by_key(self, col, stage_name: str = None):
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:160]: We need to have it lazy, d.items() is generator-like object, but for-loop is executed eagerly. What do you think about the following approach

```
def group_by_key(self, ...):
   def group_by_key_generator(...):
         d = defaultdict(list)
         for key, value in col:
             d[key].append(value)
         for item in d.items():
             yield item
   return group_by_key_generator
```

**Review by dvadym — CHANGES_REQUESTED** [review]
Thanks for contributing! I left a comment

File: pipeline_dp/pipeline_operations.py (hunk 1 above)
Comment (zachferr) [reply pipeline_dp/pipeline_operations.py:160]: I updated it, let me know what you think.
Thanks for catching this I didn't consider that

File: pipeline_dp/pipeline_operations.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import abc
 import apache_beam as beam
 import apache_beam.transforms.combiners as combiners
+from collections import defaultdict
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:8]: I recalled that by Google style guide, packages should be imported not classes. 
https://google.github.io/styleguide/pyguide.html#224-decision

Please update to 
import collections

File: pipeline_dp/pipeline_operations.py (hunk 1 above)
Comment (dvadym) [reply pipeline_dp/pipeline_operations.py:160]: Ah, yeah, thanks, () was missing in "return group_by_key_generator"

**Review by dvadym — CHANGES_REQUESTED** [review]
Thanks, I left a minor comment

**Comment by replomancer** [comment]
LGTM  (I guess only thing left is the style fix)

**Review by replomancer — APPROVED** [review]

File: pipeline_dp/pipeline_operations.py [hunk 3]
```diff
@@ -156,8 +157,14 @@ def map_tuple(self, col, fn, stage_name: str = None):
     def map_values(self, col, fn, stage_name: str):
         pass
 
-    def group_by_key(self, col, stage_name: str):
-        pass
+    def group_by_key(self, col, stage_name: str = None):
```
Comment (chinmayshah99) [inline pipeline_dp/pipeline_operations.py:160]: change this to `stage_name: Optional[str]` to confor to mypy standards.

**Review by chinmayshah99 — CHANGES_REQUESTED** [review]

File: pipeline_dp/pipeline_operations.py (hunk 3 above)
Comment (zachferr) [reply pipeline_dp/pipeline_operations.py]: Added it

**Review by dvadym — APPROVED** [review]
Thanks!

**Pull request merged**